replica/table.cc

```diff
@@ -627,7 +627,9 @@ future<> storage_group_manager::for_each_storage_group_gently(std::function<futu
 
 void storage_group_manager::for_each_storage_group(std::function<void(size_t, storage_group&)> f) const {
     for (auto& [id, sg]: _storage_groups) {
-        f(id, *sg);
+        if (auto holder = try_hold_gate(sg->async_gate())) {
```

Comment:
Is this where the check for stopped group is hidden?

Reply:
Yes, it's similar to futurized variants (they're right above this function so you can verify it, introduced in c539b7c861eb1afd6f68c102243158428b438481), so we missed this non-futurized one and resulted in this problem. If storage group gate is closed, we don't perform the operation since we consider it became unused.

Reply:
There was a wrong assumption that traversal of storage group list would always start with all groups "opened", and they could only be closed during the iteration. Like stopping the group and deallocating it is atomic, but it's not true.

Reply:
We tried making it atomic before but it failed for reasons I don't remember, so we switched to current approach.

Reply:
And we considered that a tablet that escapes the iteration is not a problem, since it's not longer part of the node at cleanup stage. Operations like major compaction are inheritently racy with migrations. We're not willing to block topology changes either for the length of long operations like major, so we're living with it by the time being.